bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerHandle.java

```diff
@@ -1898,302 +1788,81 @@ void handleBookieFailure(final Map<Integer, BookieSocketAddress> failedBookies)
-            // again when it was trying re-reading the metatada, re-kick the reread again
-            if (metadata.isNewerThan(newMeta)) {
-                if (LOG.isDebugEnabled()) {
-                    LOG.debug("[EnsembleChange-L{}-{}] : reread metadata because local metadata is newer.",
-                        new Object[]{ledgerId, ensembleChangeIdx});
-                }
-                rereadMetadata(this);
-                return true;
-            }
-            // make sure the metadata doesn't changed by other ones.
-            if (metadata.isConflictWith(newMeta)) {
-                if (LOG.isDebugEnabled()) {
-                    LOG.debug("[EnsembleChange-L{}-{}] : metadata is conflicted, local metadata = \n {} \n,"
-                        + " zk metadata = \n {} \n", ledgerId, ensembleChangeIdx, metadata, newMeta);
-                }
-                return false;
-            }
+    void ensembleChangeLoop(List<BookieSocketAddress> origEnsemble, Map<Integer, BookieSocketAddress> failedBookies) {
+        // when the ensemble changes are too frequent, close handle
+        if (numEnsembleChanges.incrementAndGet() > clientCtx.getConf().maxAllowedEnsembleChanges) {
             if (LOG.isDebugEnabled()) {
-                LOG.info("[EnsembleChange-L{}-{}] : resolved ledger metadata conflict and writing to zookeeper,"
-                        + " local meta data is \n {} \n, zk meta data is \n {}.",
-                        ledgerId, ensembleChangeIdx, metadata, newMeta);
+                LOG.debug("Ledger {} reaches max allowed ensemble change number {}",
+                          ledgerId, clientCtx.getConf().maxAllowedEnsembleChanges);
             }
-            // update znode version
-            metadata.setVersion(newMeta.getVersion());
-            // merge ensemble infos from new meta except last ensemble
-            // since they might be modified by recovery tool.
-            metadata.mergeEnsembles(newMeta.getEnsembles());
-            writeLedgerConfig(new ChangeEnsembleCb(ensembleInfo, curBlockAddCompletions,
-                    ensembleChangeIdx));
-            return true;
+            handleUnrecoverableErrorDuringAdd(WriteException);
+            return;
         }
 
-        @Override
-        public String toString() {
-            return String.format("ReReadLedgerMetadata(%d)", ledgerId);
-        }
+        new MetadataUpdateLoop(
+                clientCtx.getLedgerManager(), getId(),
+                this::getLedgerMetadata,
+                (metadata) -> {
+                    List<BookieSocketAddress> lastEnsemble = metadata.getLastEnsembleValue();
+                    boolean failedBookieInEnsemble = failedBookies.entrySet().stream()
+                        .anyMatch((e) -> lastEnsemble.get(e.getKey()).equals(e.getValue()));
+                    return !metadata.isClosed() && !metadata.isInRecovery() && failedBookieInEnsemble;
```

Comment:
```
!metadata.isClosed() &&
!metadata.isInRecovery() &&
failedBookies.entrySet().stream()
                        .anyMatch((e) -> lastEnsemble.get(e.getKey()).equals(e.getValue()))
```

so we only compute `failedBookieInEnsmeble` when metadata is open.

Reply:
changed